bookkeeper-server/src/main/java/org/apache/bookkeeper/client/TryPendingReadLacOp.java

```diff
@@ -0,0 +1,152 @@
+ * Use {@link PendingLacOp} if you need quorum coverage checking.
+ */
+
+class TryPendingReadLacOp implements ReadLacCallback {
+    static final Logger LOG = LoggerFactory.getLogger(TryPendingReadLacOp.class);
+    LedgerHandle lh;
+    LacCallback cb;
+    int numResponsesPending;
+    volatile boolean completed = false;
+    volatile boolean hasValidResponse = false;
+    int lastSeenError = BKException.Code.ReadException;
+    RecoveryData maxRecoveredData;
+    long maxLac;
+    final List<BookieSocketAddress> currentEnsemble;
+    final BookieClient bookieClient;
+
+    /*
+     * Wrapper to get Lac from the request
+     */
+    interface LacCallback {
+        void getLacComplete(int rc, long lac);
+    }
+
+    TryPendingReadLacOp(LedgerHandle lh, BookieClient bookieClient,
+                        List<BookieSocketAddress> ensemble, LacCallback cb) {
+        this.lh = lh;
+        this.cb = cb;
+        this.maxLac = lh.getLastAddConfirmed();
+        this.numResponsesPending = lh.getLedgerMetadata().getEnsembleSize();
+        this.bookieClient = bookieClient;
+        this.maxRecoveredData = new RecoveryData(maxLac, 0);
+        this.currentEnsemble = ensemble;
+    }
+
+    public void initiate() {
+        for (int i = 0; i < currentEnsemble.size(); i++) {
+            bookieClient.readLac(currentEnsemble.get(i),
+                    lh.ledgerId, this, i);
+        }
+    }
+
+    @Override
+    public void readLacComplete(int rc, long ledgerId, final ByteBuf lacBuffer, final ByteBuf lastEntryBuffer,
+            Object ctx) {
+        int bookieIndex = (Integer) ctx;
+
+        numResponsesPending--;
+
+        if (completed) {
+            return;
+        }
+
+
+        if (rc == BKException.Code.OK) {
+            try {
+                // Each bookie may have two store LAC in two places.
+                // One is in-memory copy in FileInfo and other is
+                // piggy-backed LAC on the last entry.
+                // This routine picks both of them and compares to return
+                // the latest Lac.
+
+                // lacBuffer and lastEntryBuffer are optional in the protocol.
+                // So check if they exist before processing them.
+                long newLac = LedgerHandle.INVALID_ENTRY_ID;
+                // Extract lac from FileInfo on the ledger.
+                if (lacBuffer != null && lacBuffer.readableBytes() > 0) {
+                    long lac = lh.macManager.verifyDigestAndReturnLac(lacBuffer);
+                    if (lac > maxLac) {
+                        newLac = lac;
+                    }
+                }
+                // Extract lac from last entry on the disk
+                if (lastEntryBuffer != null && lastEntryBuffer.readableBytes() > 0) {
+                    RecoveryData recoveryData = lh.macManager.verifyDigestAndReturnLastConfirmed(lastEntryBuffer);
+                    long recoveredLac = recoveryData.getLastAddConfirmed();
```

Comment:
isn't it called piggyBackedLAC?

Reply:
renamed to newLac, it is clearer, thanks